### DISCUSSION TITLE
Why server-config configured so dirty?

### DISCUSSION THREAD

**Discussion by bookvik** [mikub/titanoboa#40] [Ideas]
```
(defn init-config! [& [cfg host]]
  (if cfg
    (alter-var-root #'server-config (constantly cfg))
    (if-let [config-path (System/getProperty "boa.server.config.path")]
      (load-file config-path)
      (if-let [cp-config (io/resource "boa-server-config.clj")]
        (load-string (slurp cp-config))
        (throw (IllegalStateException. "Titanoboa server config file was not found. It should be either on classpath (as \"boa-server-config.clj\") or its path should be denoted by \"boa.server.config.path\" system property.")))))
  (alter-var-root #'server-config #(util/deep-merge (get-default-config host) %))
  (alter-var-root #'server-config assoc :node-id (get-node-id server-config))
  (alter-var-root #'server-config assoc :systems-catalogue (into {}
                                                                 (mapv (fn [[k v]]
                                                                         [k (merge v {:system-def-source (clojure.repl/source-fn (var->symbol (:system-def v)))};;FIXME this does not seem to work in uberjar
                                                                                   (when (:worker-def v) {:worker-def-source (clojure.repl/source-fn (var->symbol (:worker-def v)))}))])
                                                                       (:systems-catalogue server-config)))))
```

The settings for the rabbit did not work for me and I tried to figure out what was going wrong.

And I was a little surprised that the config function is so dirty.

Given that the config is still passed as the first parameter to the function, maybe it could have been done differently?

Well ... Just merge something where you need to =)

I began to write a test for it and I had to make a dependency on the config server

and to check the chain of calls as an init config start systems start components have to replace the systems config and run them idle to get the result (super dirty)

I understand that they are marked with an exclamation mark, but this is used in one place and it seems there is no need for this

If you do not mind, I would have thought and suggested a cleaner solution

**Comment by mikub** [comment]
@bookvik yep, definitely this could be done in a bit cleaner way :) `server-config` is probably one of few vars that are used and that are mutable, but it probably might not be entirely necessary. 
And of course the multiple mutations to it within `init-config!` fn are not much elegant :)  
Feel free to submit a PR.
